### PR TITLE
Fix #5885: generics break panache repo enhancer

### LINKED ISSUE #5885: Abstraction layer over PanacheRepositoryBase
**Describe the bug**
When implementing an abstraction over the PanacheRepositoryBase, I get the error when calling findById method:

org.hibernate.UnknownEntityTypeException: Unable to locate persister: org.acme.base.AbstractBaseLongEntity

**Expected behavior**
findById() should work when using an Abstraction layer, as findAll() just works.

**Actual behavior**
(Describe the actual behavior clearly and concisely.)

**To Reproduce**
1. clone this repo: https://github.com/lssoares/panache-abstraction-problem
2. run ./mvn clean package && mvn quarkus:dev 
3. Use swagger to execute the following webservice: http://0.0.0.0:8080/example/1

**Additional context**
1.Calling the endpoint that uses findAll() works fine: http://0.0.0.0:8080/example    
2. If you edit ExampleRepository.java and swap the lines indicated there, both endpoints will work.

### PR DESCRIPTION
**Pull request by FroMage**
Do not enhance generic/abstract repos
Use proper generics resolution when looking up the repository type
arguments

Fixes: #5885

### CODE REVIEW

**Comment by FroMage** [comment]
@loicmathieu it's very possible you have the same bug in mongo-panache, and will need the same fix (do not enhance abstract/generic repos).

It's also possible that you have the same issue in your reactive-mongo-panache repo IF your entity base type is generic.

**Review by geoand — APPROVED** [review]
:+1:

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java [hunk 1]
```diff
@@ -122,4 +128,11 @@ void build(CombinedIndexBuildItem index,
         }
     }
 
+    private boolean skipRepository(ClassInfo classInfo) {
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java]: Maybe rename to `isNotLeafType` and put it in the common processor ?

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java [hunk 2]
```diff
@@ -90,15 +90,13 @@ private String recursivelyFindEntityTypeFromClass(DotName clazz, DotName reposit
...
-            final ClassInfo classByName = indexView.getClassByName(clazz);
-            for (org.jboss.jandex.Type type : classByName.interfaceTypes()) {
-                if (type.name().equals(repositoryDotName)) {
-                    org.jboss.jandex.Type entityType = type.asParameterizedType().arguments().get(0);
-                    return entityType.name().toString().replace('.', '/');
-                }
-            }
-
-            return recursivelyFindEntityTypeFromClass(classByName.superName(), repositoryDotName);
+            List<org.jboss.jandex.Type> typeParameters = io.quarkus.deployment.util.JandexUtil
```
Comment (loicmathieu) [inline extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java]: `io.quarkus.deployment.util.JandexUtil` maybe use import statement ?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
`mongodb-panache` and the future `mongodb-panache-reactive` suffers from the same bug.
If you don't have time to fix it in `mongodb-panache`inside the same PR, can you open an issue for it ?

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java (hunk 2 above)
Comment (FroMage) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java]: Well there are more than one…

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java (hunk 1 above)
Comment (FroMage) [reply extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java]: I can pull it up, but I wanted the name to be clear that this is only for repositories.

**Comment by FroMage** [comment]
> If you don't have time to fix it in `mongodb-panache` inside the same PR, can you open an issue for it ?

I can do it in this PR, should be simple.

File: extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java (hunk 2 above)
Comment (loicmathieu) [reply extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheRepositoryEnhancer.java]: There are more than one `JandexUtil` used inside the enhancer ? Stange ...

File: extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java (hunk 1 above)
Comment (FroMage) [reply extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheResourceProcessor.java]: moved up.

**Comment by FroMage** [comment]
Added test/fix for mongo too.

**Review by loicmathieu — APPROVED** [review]
LGTM

**Comment by FroMage** [comment]
Huh, so native test fails, due to who knows what, but I am already rebased on master.

**Comment by FroMage** [comment]
It passed after restart

**Comment by MartinX3** [comment]
@FroMage Here we have the problem, that a Repository listAll() call gives us every DataBase entry in every table.
As example we made a trainer and the call for courses gives us all trainer, too.

`abstract class BaseRepository<T>(@Inject val entityManager: EntityManager) : PanacheRepositoryBase<T, UUID> {`

`@Dependent
class TrainerRepository(entityManager: EntityManager) : BaseRepository<Trainer>(entityManager) {`

`@Dependent
class CourseRepository(entityManager: EntityManager) : BaseRepository<Course>(entityManager) {`

**Comment by FroMage** [comment]
@MartinX3 is this with the current master which has the fix? I think it's the same bug and it should be fixed.

**Comment by MartinX3** [comment]
@FroMage Thank you for your fast response.
We use the maven 1.1.0.final release in our company.
Nice to hear that it will get fixed :)
Will it be released as 1.1.1 or do we need to wait for 1.2.0 (which has no ETA?)

Thank you all for the fix. :)

**Comment by FroMage** [comment]
That's a question for @gsmet but IIRC 1.2 is planned for January.

**Pull request merged**